### PR TITLE
PWA chapter

### LINKED ISSUE #164: Write content: Chapter 11. PWA
Part | Chapter | Authors | Reviewers | Analysts | Metrics | Analysis
-- | -- | -- | -- | -- | -- | --
II. User Experience | 11. PWA | @tomayac @jeffposnick | @hyperpress | @jrharalson | #13 | #92

### Due by: October 7

- [x] Review the [results spreadsheet](https://docs.google.com/spreadsheets/d/19BI3RQc_vR9bUPPZfVsF_4gpFWLNT6P0pLcAdL-A56c/edit?usp=sharing) and work with your analyst(s) (@jrharalson) if you have any questions about the data.
- [x] Write the contents of your chapter and get it peer reviewed (@hyperpress).
- [ ] File a PR to merge your written contents in markdown format to the [src/content/en/2019](https://github.com/HTTPArchive/almanac.httparchive.org/tree/master/src/content/en/2019) directory.

Refer to the [Authors' Guide](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Authors'-Guide) for writing advice.

### PR DESCRIPTION
**Pull request by rviscomi**
Closes #164 

cc @jeffposnick @tomayac

### CODE REVIEW

**Comment by tomayac** [comment]
As announced over email, I’m OoO the next two weeks and hope @jeffposnick can help move this over the finish line.

File: src/content/en/2019/pwa.md [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+authors: [tomayac, jeffposnick]
+reviewers: [hyperpress]
+translators: []
+---
+
+## Introduction
+Progressive Web Apps (PWA) are a new class of web applications, building on top of platform primitives 
+like the [Service Worker APIs](https://developer.mozilla.org/en/docs/Web/API/Service_Worker_API). 
+Service workers allow apps to support network-independent loading by acting as a network proxy, 
+intercepting your web app's outgoing requests and replying with programmatic or cached responses.
```
Comment (ahmadawais) [inline src/content/en/2019/pwa.md]:
```suggestion
intercepting your web app's outgoing requests, and replying with programmatic or cached responses. 
```

File: src/content/en/2019/pwa.md [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+allow users to install PWAs to their devices’ home screens.
+
+Service workers were [first implemented in Chrome 40](https://blog.chromium.org/2014/12/chrome-40-beta-powerful-offline-and.html), 
+back in December 2014, and the term Progressive Web Apps was 
+[coined by Frances Berriman and Alex Russell](https://infrequently.org/2015/06/progressive-apps-escaping-tabs-without-losing-our-soul/) 
+in 2015. As service workers are now finally [implemented in all major browsers](https://jakearchibald.github.io/isserviceworkerready/), 
+we were wondering how many PWAs are actually out there, and how do they make use of these new technologies? 
+Certain advanced APIs like [Background Sync](https://developers.google.com/web/updates/2015/12/background-sync) 
+are currently still [only available on Chromium-based browsers](https://caniuse.com/#feat=background-sync), 
+so as an additional question we looked into which features these PWAs actually use.
```
Comment (ahmadawais) [inline src/content/en/2019/pwa.md]:
```suggestion
so as an additional question, we looked into which features these PWAs actually use.

```

File: src/content/en/2019/pwa.md [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+they are partitioned, but multiple associated tables can be queried using the [wildcard symbol '*'](https://cloud.google.com/bigquery/docs/querying-wildcard-tables).
+For our research, three families of tables are relevant:
+
+- `httparchive.lighthouse.*`, which contains data about [Lighthouse](https://developers.google.com/web/tools/lighthouse/) runs. Note that Lighthouse data is only available for mobile pages.
+- `httparchive.pages.*`, which contain the JSON-encoded parent documents’ HAR data.
+- `httparchive.response_bodies.*`, which contains the raw response bodies of all resources and sub-resources of all sites in the archive.
+
+## Service Workers
+### Service Worker Registrations and Installability
+The first metric we explore are service worker installations. Looking at the data exposed through
```
Comment (ahmadawais) [inline src/content/en/2019/pwa.md]:
```suggestion
The first metric we explore is service worker installations. Looking at the data exposed through 
```

File: src/content/en/2019/pwa.md [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+- `httparchive.pages.*`, which contain the JSON-encoded parent documents’ HAR data.
+- `httparchive.response_bodies.*`, which contains the raw response bodies of all resources and sub-resources of all sites in the archive.
+
+## Service Workers
+### Service Worker Registrations and Installability
+The first metric we explore are service worker installations. Looking at the data exposed through 
+feature counters in the HTTP Archive, we find that 0.44% of all desktop and 0.37% of all mobile pages 
+register a service worker, and both curves over time are steeply growing. Now this might not look overly 
+impressive, but taking traffic data from Chrome Platform Status into account, we can see that about 
+[15% of all page loads were controlled by a service worker](https://www.chromestatus.com/metrics/feature/timeline/popularity/990),
```
Comment (ahmadawais) [inline src/content/en/2019/pwa.md]:
```suggestion
[a service worker controlled 15% of all page loads](https://www.chromestatus.com/metrics/feature/timeline/popularity/990),
```

File: src/content/en/2019/pwa.md [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+**Figure 1:** Service Worker installation over time for desktop and mobile
+
+Looking at Lighthouse data in the HTTP Archive, 1.56% of mobile pages are [installable](https://developers.google.com/web/tools/lighthouse/audits/install-prompt),
+that is, they pass Lighthouse’s *user can be prompted to install the web app* test. 
+Lighthouse tests currently are only available for mobile pages. To control the install experience, 
+0.82% of all desktop and 0.94% of all mobile pages use the [`OnBeforeInstallPrompt` interface](https://w3c.github.io/manifest/#beforeinstallpromptevent-interface).
+
+### Service Worker Events
+
+In a service worker one can [listen for a number of events](https://developers.google.com/web/fundamentals/primers/service-workers/lifecycle):
```
Comment (ahmadawais) [inline src/content/en/2019/pwa.md]:
```suggestion
In a service worker, one can [listen for many events](https://developers.google.com/web/fundamentals/primers/service-workers/lifecycle):
```

File: src/content/en/2019/pwa.md [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+- `notificationclose`, which occurs when a notification is being closed.
+`- message`, which occurs when a message sent via `postMessage()`  arrives.
+- `sync`, which occurs when a Background Sync event occurs.
+
+We have examined which of these events are being listened to by service workers we could find in the HTTP Archive. 
+The results for mobile and desktop are very similar with `fetch`, `install`, and `activate` being the three 
+most popular events, followed by `notificationclick` and `push`. If we interpret these results, offline use 
+cases that service workers enable are the most attractive feature for app developers, far ahead of 
+push notifications. Due to its limited availability, and less common use case, background sync doesn’t 
+play a big role at the moment.
```
Comment (ahmadawais) [inline src/content/en/2019/pwa.md]:
```suggestion
play a significant role at the moment. 
```

File: src/content/en/2019/pwa.md [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+<bar chart of 11_03 mobile>
+
+**Figure 2a:** Service worker events on mobile, ordered by decreasing frequency.
+
+<bar chart of 11_03 desktop>
+
+**Figure 2b:** Service worker events on desktop, ordered by decreasing frequency.
+
+### Service Worker File Sizes
+File size or lines of code are in general a bad proxy for the complexity of the task at hand.
```
Comment (ahmadawais) [inline src/content/en/2019/pwa.md]:
```suggestion
File size or lines of code are, in general, a bad proxy for the complexity of the task at hand. 
```

File: src/content/en/2019/pwa.md [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+**Figure 2a:** Service worker events on mobile, ordered by decreasing frequency.
+
+<bar chart of 11_03 desktop>
+
+**Figure 2b:** Service worker events on desktop, ordered by decreasing frequency.
+
+### Service Worker File Sizes
+File size or lines of code are in general a bad proxy for the complexity of the task at hand. 
+In this case, however, it is definitely interesting to compare (compressed) file sizes of service workers 
+for mobile and desktop. The median service worker file on desktop is 895 bytes, whereas on mobile it’s 694 bytes.
```
Comment (ahmadawais) [inline src/content/en/2019/pwa.md]:
```suggestion
for mobile and desktop. The median service worker file on desktop is 895 bytes, whereas, on mobile, it’s 694 bytes. 
```

File: src/content/en/2019/pwa.md [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+`icons`, `screenshots`, `categories`, `iarc_rating_id`, `start_url`, `display`, `orientation`, 
+`theme_color`, `background_color`, `scope`, `serviceworker`, `related_applications`, and 
+`prefer_related_applications`. The only property that we didn’t observe in the wild was 
+`iarc_rating_id`, which is a string that represents the International Age Rating Coalition (IARC) 
+certification code of the web application. It is intended to be used to determine which ages 
+the web application is appropriate for. The proprietary properties we encountered still 
+frequently were `gcm_sender_id` and `gcm_user_visible_only` from the legacy 
+Google Cloud Messaging (GCM) service. Interestingly there’re almost no differences between mobile 
+and desktop. On both platforms, however, there’s a long tail of properties that are not interpreted 
+by browsers but that contain potentially useful metadata like `author` or `version`. We also found
```
Comment (ahmadawais) [inline src/content/en/2019/pwa.md]:
```suggestion
by browsers, but that contains potentially useful metadata like `author` or `version`. We also found 
```

File: src/content/en/2019/pwa.md [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+
+**Figure 4a:** Web App Manifest properties ordered by decreasing popularity on mobile.
+
+<bar chart of 11_04 mobile>
+
+**Figure 4b:** Web App Manifest properties ordered by decreasing popularity on desktop.
+
+### Display Values
+Looking at the values developers set for the `display` property, it becomes immediately clear 
+that they want PWAs to be perceived as “proper” apps that don’t  reveal their web technology origins.
```
Comment (ahmadawais) [inline src/content/en/2019/pwa.md]:
```suggestion
that they want PWAs to be perceived as “proper” apps that don’t reveal their web technology origins. 
```

File: src/content/en/2019/pwa.md [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+**Figure 4a:** Web App Manifest properties ordered by decreasing popularity on mobile.
+
+<bar chart of 11_04 mobile>
+
+**Figure 4b:** Web App Manifest properties ordered by decreasing popularity on desktop.
+
+### Display Values
+Looking at the values developers set for the `display` property, it becomes immediately clear 
+that they want PWAs to be perceived as “proper” apps that don’t  reveal their web technology origins. 
+By choosing `"standalone"` they make sure no browser UI is shown to the end user. This is reflected
```
Comment (ahmadawais) [inline src/content/en/2019/pwa.md]:
```suggestion
By choosing `"standalone"`, they make sure no browser UI is shown to the end-user. This is reflected 
```

File: src/content/en/2019/pwa.md [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+
+<bar chart of 11_04 mobile>
+
+**Figure 4b:** Web App Manifest properties ordered by decreasing popularity on desktop.
+
+### Display Values
+Looking at the values developers set for the `display` property, it becomes immediately clear 
+that they want PWAs to be perceived as “proper” apps that don’t  reveal their web technology origins. 
+By choosing `"standalone"` they make sure no browser UI is shown to the end user. This is reflected 
+by the majority of apps that make use of the `prefers_related_applications` property: more that 97%
```
Comment (ahmadawais) [inline src/content/en/2019/pwa.md]:
```suggestion
by the majority of apps that make use of the `prefers_related_applications` property: more than 97% 
```

File: src/content/en/2019/pwa.md [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+
+**Figure 5a:** Values for the `display` property on mobile.
+
+<11_04c desktop>
+
+**Figure 5b:** Values for the `display` property on desktop.
+
+### Category Values
+The `categories` member describes the expected application categories to which the web application belongs. 
+It is only meant as a hint to catalogs or stores listing web applications and it is expected that
```
Comment (ahmadawais) [inline src/content/en/2019/pwa.md]:
```suggestion
It is only meant as a hint to catalogs or stores listing web applications, and it is expected that 
```

File: src/content/en/2019/pwa.md [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+
+**Figure 8a:** Popular orientation values on mobile.
+
+<11_04g desktop>
+
+**Figure 8b:** Popular orientation values on desktop.
+
+## Workbox
+Workbox is a set of libraries that help with common service worker use cases. For instance, Workbox has tools 
+that can plug in to your build process and generate a manifest of files, which are then precached by your
```
Comment (ahmadawais) [inline src/content/en/2019/pwa.md]:
```suggestion
that can plug into your build process and generate a manifest of files, which are then precached by your 
```

File: src/content/en/2019/pwa.md [hunk 15]
```diff
@@ -0,0 +1,208 @@
+---
+part_number: [II]
+chapter_number: [11]
+title: Progressive Web Apps
+authors: [tomayac, jeffposnick]
+reviewers: [hyperpress]
```
Comment (ahmadawais) [inline src/content/en/2019/pwa.md]:
```suggestion
reviewers: [hyperpress, ahmadawais]
```

Comment (ahmadawais) [reply src/content/en/2019/pwa.md]: @rviscomi just finished reviewing the PWA chapter, figured I should help out while waiting for the Media chapter. I hope it's fine to add my name in the reviewers here.

Comment (rviscomi) [reply src/content/en/2019/pwa.md]: Absolutely, thank you for reviewing!

**Comment by rviscomi** [comment]
> A couple of minor nitpicks mostly grammatical and such.

Thanks @ahmadawais! To confirm, you've also given it a review for technical correctness?

**Comment by ahmadawais** [comment]
Yes, to the best of my knowledge. I don't see the images in there. But it was an interesting read. 👍

**Comment by rviscomi** [comment]
Ah that's right. This PR is a markdown version of the doc authored by Tom and Jeff: https://docs.google.com/document/d/1MOt6doqXdL1yrvZvpRIDv37ulFpUBBj1cxLi_l8k1gQ/edit?usp=sharing

You'll need to request view access though.

**Comment by ahmadawais** [comment]
Yes, I just requested access. Do I need to do the same for the Media chapter? Can you share its doc link? And yes, I have already checked the `Web Almanac: PWA Analysis` sheet.

**Comment by tomayac** [comment]
Access to the document was granted, thanks @ahmadawais.

**Comment by rviscomi** [comment]
Great, thanks again @ahmadawais. Leaving this PR open for @hyperpress to review.

**Comment by hyperpress** [comment]
Will review after work. Thanks.

**Comment by hyperpress** [comment]
Full review. PWA information is logically presented and accurate. Great work that users will no doubt find instructive.

**Pull request merged**